src/pyodide/internal/workers.py

```diff
@@ -149,54 +157,202 @@ def json(
             raise _to_python_exception(exc) from exc
 
 
-# TODO: Implement pythonic blob API
-FormDataValue = "str | js.Blob"
+FormDataValue = "str | js.Blob | Blob"
+
+
+def _py_value_to_js(item: FormDataValue) -> "str | js.Blob":
+    if isinstance(item, Blob):
+        return item.native
+    else:
+        return item
+
+
+def _js_value_to_py(item: FormDataValue) -> "str | Blob":
+    if hasattr(item, "constructor") and (item.constructor.name in ("Blob", "File")):
+        return Blob(item)
+    else:
+        return item
 
 
 class FormData(MutableMapping[str, FormDataValue]):
-    def __init__(self, form_data: "js.FormData | None | dict[str, FormDataValue]"):
+    """
+    This API follows that of https://pypi.org/project/multidict/.
+    """
+
+    def __init__(
+        self, form_data: "js.FormData | None | dict[str, FormDataValue]" = None
+    ):
         if form_data:
             if isinstance(form_data, dict):
-                self.js_form_data = js.FormData.new()
-                for item in form_data.items():
-                    self.js_form_data.append(item[0], item[1])
+                self._js_form_data = js.FormData.new()
+                for k, v in form_data.items():
+                    self._js_form_data.append(k, _py_value_to_js(v))
             else:
-                self.js_form_data = form_data
+                self._js_form_data = form_data
         else:
-            self.js_form_data = js.FormData.new()
+            self._js_form_data = js.FormData.new()
 
-    def __getitem__(self, key: str) -> list[FormDataValue]:
-        return list(self.js_form_data.getAll(key))
+    def __getitem__(self, key: str) -> FormDataValue:
+        return _js_value_to_py(self._js_form_data.get(key))
 
-    def __setitem__(self, key: str, value: list[FormDataValue]):
-        self.js_form_data.delete(key)
-        for item in value:
-            self.js_form_data.append(key, item)
+    def __setitem__(self, key: str, value: FormDataValue):
+        if isinstance(value, list):
+            raise TypeError("Expected single item in arguments to FormData.__setitem__")
+        self._js_form_data.set(key, _py_value_to_js(value))
 
-    def append(self, key: str, value: FormDataValue):
-        self.js_form_data.append(key, value)
+    def append(self, key: str, value: FormDataValue, filename: str | None = None):
+        if filename:
+            self._js_form_data.append(key, _py_value_to_js(value), filename)
+        else:
+            self._js_form_data.append(key, _py_value_to_js(value))
```

Comment:
These can be unified:
```suggestion
        self._js_form_data.append(key, _py_value_to_js(value), filename)
```